### PR TITLE
refactor: extract getRecommendedPosts into src/lib

### LINKED ISSUE #639: Extract recommended-posts algorithm out of [slug].astro
## Problem

A non-trivial tag-based similarity scoring algorithm lives directly inside `getStaticPaths` in `src/pages/[slug].astro` (lines 22–47). It builds a tag index, accumulates scores across shared tags, tiebreaks by date, and takes the top 6. It is untested, invisible to search unless you already know to look in a page file, and cannot be imported or called from anywhere else.

## Solution

Extract to a `getRecommendedPosts(post, allPosts)` module in `src/lib/`. The interface is small (a post + a collection → a ranked list); the implementation is the scoring logic. Pages just call it.

## Benefits

- **Locality** — the scoring algorithm is in one place with a name.
- **Testability** — ranking rules can be tested directly (e.g. "prefers shared-tag count over recency, all else equal").
- **AI-navigability** — the algorithm is findable by name.

### PR DESCRIPTION
**Pull request by narthur**
## Summary

Moves the tag-based similarity scoring algorithm out of `getStaticPaths` in `src/pages/[slug].astro` into a named, testable, importable `src/lib/getRecommendedPosts.ts` module. The page now just calls `getRecommendedPosts(post, publishedPosts)` per post.

Behavior is preserved exactly:
- Same scoring (count of shared tags)
- Same tiebreak (most recent date wins)
- Same cap of 6 recommendations
- Same self-exclusion

## Test plan

- [x] New colocated unit tests in `src/lib/getRecommendedPosts.spec.ts` cover: shared-tag preference, date tiebreak, 6-post cap, self-exclusion, no-shared-tags edge case (5 tests, all passing)
- [x] `pnpm vitest run` — full test suite passes (104/104)
- [x] `pnpm check:ts` — typecheck passes
- [x] `pnpm lint` — no new lint errors

Note: `pnpm test:snapshot` fails locally because `ETHERPAD_DOMAIN` is not set in this worktree's environment; this is unrelated to this change.

Closes #639

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/pages/[slug].astro [hunk 1]
```diff
@@ -19,31 +20,9 @@ export async function getStaticPaths() {
...
-        }
-      }
-    }
-    const sorted = [...scores.entries()]
-      .sort((a, b) => b[1] - a[1] || slugToPost.get(b[0])!.date.getTime() - slugToPost.get(a[0])!.date.getTime())
-      .slice(0, 6)
-      .map(([slug]) => slugToPost.get(slug)!);
-    recommendedMap.set(post.slug, sorted);
+    recommendedMap.set(post.slug, getRecommendedPosts(post, publishedPosts));
   }
```
[the remark this answers is not in the document]
Comment (narthur) [reply src/pages/[slug].astro:26]: Fixed by memoizing the tag index + slug map per `allPosts` reference via a WeakMap. Repeat calls with the same `allPosts` array now reuse the index, restoring the original O(N) build behavior. Signature unchanged.

File: src/lib/getRecommendedPosts.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  const tagIndex = new Map<string, Post[]>();
+  for (const p of allPosts) {
+    for (const tag of p.tags) {
+      const list = tagIndex.get(tag);
+      if (list) list.push(p);
+      else tagIndex.set(tag, [p]);
+    }
+  }
+
+  const slugToPost = new Map(allPosts.map((p) => [p.slug, p]));
```
[the remark this answers is not in the document]
Comment (narthur) [reply src/lib/getRecommendedPosts.ts:27]: Addressed via WeakMap-keyed memoization on `allPosts` — the indexing now runs once per unique array, not per call. See the updated `buildIndex` helper.
